articles/ci_pipeline_and_custom_tools_of_android_projects.md

```diff
@@ -0,0 +1,464 @@
+
+#### Why Bash and not Kotlin Script, etc?
+
+It's fair to say that [Bash][bash] is far, [very far][bash tutorial] from a perfect scripting language. It has lots of caveats and strangies.
+
+However, there are several reasons why we chose it and continued to maintain CI scripts in Bash:
+
+- Availability: We can't launch Docker container directly because we have some pre-setup, so we need a language that runs on most Linux distros out of the box.
+- Compatibility: it's pretty easy to write Bash scripts that will work on even relatively old versions of Bash interpreter.
+- Usability: Mostly our build scripts check environment variables, concatenate strings and run other processes. This is what Bash was designed for!
+- Complexity: If we need to do more complex stuff, like [creating and running Android Emulators in parallel][swarmer], we write it in Kotlin (see [Swarmer][swarmer]) and then invoke it from Bash.
+
+We've tried to migrate some of the scripts to [Kotlin Script][kotlin script], but immediately faced issues:
+
+- Java APIs for launching external processes ~~suck~~ are not very great, even with Kotlin it's painful.
+    - Btw we have a library for that: [Commander][commander]
+- To run Kotlin Script you need JVM and Kotlin Compiler on a machine
+    - Docker may help here, but what if you need a script to start Docker Container in the first place?
+- Kotlin Script still [does not have official solution for dependency management][kotlin script depmngmt], even in small scripts you might want to use some third-party library and it's hard to do.
+- Bash tends to be faster than Kotlin Script if script is relatively small and not complex:
+    - To run `.kts` you need to compile it first, Kotlin Compiler is pretty heavy so it may take a few seconds to just start it.
+    - To *actually* run `.kts` you need to launch JVM, which is heavier than Bash interpreter.
+    - However if you're doing complex stuff and script runs for a long time, JVM will most likely beat Bash. 
+    - However however if you're planning to do complex stuff in a script, you might probably want to extract it in a separate tool and invoke from Bash.
+
+## Back to "What actually happens on CI"
+
+So now that we know about Docker and Build Scripts we can describe how we build our project:
+
+1. CI Pipeline detects change in Source Code Repository
+1. CI notifies Source Code Repository about started build
+1. CI runs `bash ci/build.sh`
+1. `ci/build.sh` prepares Docker Image
+     - Either it is already on machine
+     - Or it can be pulled from the Docker registry 
+     - Or it should be built and pushed to the registry
+1. Docker Container invokes `bash ci/internal/build.sh`
+    - If it's "Build" stage of Pipeline:
+        1. `ci/build.sh` starts Docker Container with build workspace mounted as a volume and `bash ci/internal/build.sh` as a target command
```

Comment:
Hmm, is it really child step for `Docker Container invokes bash ci/internal/build.sh`?